pps_mw_training/pipelines/cloud_base/training_data.py

```diff
@@ -0,0 +1,144 @@
+
+
+def _load_data(
+    data_files: np.ndarray,
+    input_parameters: str,
+    label_parameters: str,
+    fill_value_input: float,
+    fill_value_label: float,
+) -> list[np.ndarray]:
+    """Load, scale, and filter data."""
+    with xr.open_mfdataset(
+        [f.decode("utf-8") for f in data_files],
+        combine="nested",
+        concat_dim="nscene",
+    ) as all_data:
+
+        input_data = scale_data(all_data, input_parameters, fill_value_input)
+        label_data = scale_data(all_data, label_parameters, fill_value_label)
+        return [
+            input_data,
+            label_data,
+        ]
+
+
+def scale_data(
+    data: xr.Dataset, parameters: str, fill_value: float
+) -> np.ndarray:
+    """Scale data"""
+    params = json.loads(parameters)
+    scaler = get_scaler(params)
+    data = np.stack(
+        [
+            scaler.apply(data[p["name"]].values, idx)
+            for idx, p in enumerate(params)
+        ],
+        axis=3,
+        dtype=np.float32,
+    )
+    data[~np.isfinite(data)] = fill_value
+    return data
+
+
+@tf.function(
+    input_signature=(
+        tf.TensorSpec(shape=(None,), dtype=tf.string),
+        tf.TensorSpec(shape=(), dtype=tf.string),
+        tf.TensorSpec(shape=(), dtype=tf.string),
+        tf.TensorSpec(shape=(), dtype=tf.float32),
+        tf.TensorSpec(shape=(), dtype=tf.float32),
+    )
+)
+def load_data(
+    files,
+    input_params,
+    label_params,
+    fill_value_input,
+    fill_value_label,
+):
+    """Load netcdf dataset."""
+    return tf.numpy_function(
+        func=_load_data,
+        inp=[
+            files,
+            input_params,
+            label_params,
+            fill_value_input,
+            fill_value_label,
+        ],
+        Tout=[tf.float32, tf.float32],
+    )
+
+
+def _get_training_dataset(
+    files: list[Path],
+    batch_size: int,
+    input_parameters: list[dict[str, str | float]],
+    label_parameters: list[dict[str, str | float]],
+    fill_value_input: float,
+    fill_value_label: float,
+) -> tf.data.Dataset:
+    """Get training dataset."""
+
+    ds = tf.data.Dataset.from_tensor_slices([f.as_posix() for f in files])
+    ds = ds.batch(batch_size)
+    input_params = json.dumps(input_parameters)
+    label_params = json.dumps(label_parameters)
+    ds = ds.map(
+        lambda x: load_data(
+            x,
+            tf.constant(input_params),
+            tf.constant(label_params),
+            tf.constant(fill_value_input),
+            tf.constant(fill_value_label),
+        ),
+        num_parallel_calls=1,
+    )
+    return ds
+
+
+def get_training_dataset(
+    training_data_path: Path,
+    train_fraction: float,
+    validation_fraction: float,
+    test_fraction: float,
+    batch_size: int,
+    input_parameters: list[dict[str, str | float]],
+    label_parameters: list[dict[str, str | float]],
+    fill_value_input: float,
+    fill_value_label: float,
+    file_limit: Optional[int],
```

Comment:
Yes I thought the same. But strangely it works without that and mypy doesn't complain. I will fix it anyhow.

Reply:
Or is it because I have in train.py as parser argument:   add_file_limit: bool = False. can we have None and False togehter?

Reply:
It works fine for you since you provide the input parameter so that is why mypy does not complain. But when you have an optional parameter it makes sense to set the default value to None, otherwise a user need to supply the input parameter even if he wants to use the None option.

Reply:
and to be clear, you can not have None and False  together, but that we do not have in this case.